src/main/java/uk/sky/cirrus/locking/Lock.java

```diff
@@ -0,0 +1,174 @@
+
+        while (!this.released) {
+            try {
+                ResultSet resultSet = session.execute(query);
+                this.released = handleLockRelease(resultSet.one(), isRetryAfterWriteTimeout);
+            } catch (WriteTimeoutException wte) {
+                log.warn("Query to release lock for {} failed to execute: {}", client, wte.getMessage());
+                waitToRetryRelease();
+                isRetryAfterWriteTimeout = true;
+            } catch (Exception e) {
+                log.error("Query to release lock failed to execute for {} by client {}", name, client, e);
+                throw new CannotReleaseLockException("Query failed to execute", e);
+            }
+        }
+    }
+
+    private boolean handleLockRelease(final Row result, boolean isRetryAfterWriteTimeout) {
+
+        boolean isApplied = result.getBool("[applied]");
+
+        if (isApplied || !result.getColumnDefinitions().contains("client")) {
+            log.info("Lock released for {} by client {} at: {}", name, client, System.currentTimeMillis());
+            return true;
+        }
+
+        final UUID clientReleasingLock = result.getUUID("client");
+        if (!clientReleasingLock.equals(this.client)) {
+            if (isRetryAfterWriteTimeout) {
+                log.info("Released lock for client {} in retry attempt after WriteTimeoutException", clientReleasingLock);
+                return true;
+            } else {
+                log.error("Lock attempted to be released by a non lock holder {}", clientReleasingLock);
+                throw new IllegalStateException("Lock attempted to be released by a non lock holder: " + clientReleasingLock);
+            }
+        }
+
+        return false;
+    }
+
+    private static ResultSet tryAcquire(Session session, UUID client, Statement query) {
+        try {
+            return session.execute(query);
+        } catch (WriteTimeoutException wte) {
+            throw wte;
+        } catch (DriverException de) {
+            log.warn("Query to acquire lock for {} failed to execute", client, de);
+            throw new CannotAcquireLockException("Query failed to execute", de);
+        }
+    }
+
+    private static void waitToAcquire(LockConfig lockConfig, UUID clientId, int acquireAttempts, long startTime, UUID currentLockHolder) {
+        log.debug("Lock currently in use by client: {}", currentLockHolder);
+
+        if (timedOut(lockConfig.getTimeout(), startTime)) {
+            log.warn("Unable to acquire lock for {} after {} attempts, currently in use by client: {}, time tried: {}", clientId, acquireAttempts, currentLockHolder, (System.currentTimeMillis() - startTime));
+            throw new CannotAcquireLockException("Lock currently in use by client: " + currentLockHolder);
+        }
+
+        try {
+            Thread.sleep(lockConfig.getPollingInterval().toMillis());
+        } catch (InterruptedException e) {
+            log.debug("Lock {} was interrupted", clientId);
```

Comment:
This is not expected to happen so why don't we do it as Info so we know it has happened? Debug logging will rarely ever be on